### PR TITLE
Enable hybrid tensor + expert + data parallelism in mcore inference

### PR DESCRIPTION
**Pull request by sidsingh-nvidia**
# What does this PR do ?
This PR enabled hybrid expert + data parallelism in mcore inference w/ the coordinator. EP ranks with 0 work run dummy forward passes if any of their peers have work. Proper handling of stop signals is also done. 
Also added experimental tensor parallel support for non-expert parameters.

Todo: Test for correctness of pause/suspend. 

:warning: For major changes (either in lines of code or in its impact), please make sure to first share discuss a design-doc with the team.  

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Comment by kvareddy** [comment]
@wdykas @santhnm2 @lmcafee-nvidia  can you please take a look at this MR?

File: megatron/core/inference/engines/dynamic_engine.py [hunk 1]
```diff
@@ -1434,6 +1429,45 @@ async def run_engine(
...
+        range_push("_ep_group_has_work")
+        
+        is_stopped = self.stopped.is_set() or self.received_stop 
+        is_paused = self.paused.is_set() or self.received_pause
+        is_suspended = self.suspend_signal or self.is_suspended
+        if is_stopped or is_paused or is_suspended:
+            # attempt to pause/stop/suspend 
+            # contribute 0 work to the all-reduce 
+            # Signals are received asynchronously so some EP ranks might not have 
+            # received them yet. So in that case they might nominate +ve work
```
Comment (santhnm2) [inline megatron/core/inference/engines/dynamic_engine.py]: What is +ve?

Comment (sidsingh-nvidia) [reply megatron/core/inference/engines/dynamic_engine.py]: I should reword this. 
EP ranks can get stop/pause signals in different engine steps, in which case you want to avoid  pausing/stopping a rank until all of it's peers have received the signal. 
To make this possible, I am making EP ranks that have received a pause/stop/suspend signal contribute a `0` to this all-reduce. If even one of them hasn't received the signal they will contribute a non-zero to this all-reduce and all ranks will chug along for another engine step.

**Comment by sidsingh-nvidia** [comment]
@santhnm2 Will wait for us to resolve zmq-related flakiness before I add tests

File: megatron/core/inference/engines/dynamic_engine.py [hunk 2]
```diff
@@ -1439,6 +1439,45 @@ async def run_engine(
...
+            # So this is an *attempt* to process the signal. This rank has received the signal
+            # and passes 0 to the all-reduce. If any other rank in the EP group has not received the signal yet,
+            # it will pass a non-zero value to the all-reduce, and hence the global work will be non-zero,
+            # and we will defer processing the signal.
+            # When all ranks receive the signal, global work will be zero, and we can process the signal safely.
+            local_work = 0
+        else:
+            local_work = self.context.get_active_request_count() + len(self.waiting_request_ids)
+
+        if parallel_state.get_expert_model_parallel_world_size() > 1:
```
Comment (wdykas) [inline megatron/core/inference/engines/dynamic_engine.py]: This is urgent but in the future once https://github.com/NVIDIA/Megatron-LM/pull/2590/files goes in we will need to change this to not use global parallel state since it can change with RL

Comment (sidsingh-nvidia) [reply megatron/core/inference/engines/dynamic_engine.py]: Did you mean `not urgent`? Either way I will add a comment here

File: megatron/core/inference/engines/dynamic_engine.py [hunk 3]
```diff
@@ -1462,26 +1501,35 @@ async def run_engine_with_coordinator(
                 # needed to send one message on an IPC socket. However
                 # just to be safe, we use 20ms here.
 
+                pending_requests = self.context.get_active_request_count() + len(
+                    self.waiting_request_ids
+                )
+                ep_group_has_work = self._ep_group_has_work()
```
Comment (wdykas) [inline megatron/core/inference/engines/dynamic_engine.py]: I see that the behavior is correct even with no EP but just wondering if the stuff inside this function has any overhead? if it does we could just put it in an if only if we have ep

**Comment by wdykas** [comment]
few comments but LGTM

File: megatron/core/inference/engines/dynamic_engine.py (hunk 3 above)
Comment (sidsingh-nvidia) [reply megatron/core/inference/engines/dynamic_engine.py]: The if condition in the function skips the all-reduce if EP is 1. Outside of the all-reduce, I think the function is pretty lean.

File: megatron/core/inference/engines/dynamic_engine.py (hunk 1 above)
Comment (sidsingh-nvidia) [reply megatron/core/inference/engines/dynamic_engine.py]: Reworded for clarity.

File: megatron/core/inference/engines/dynamic_engine.py [hunk 4]
```diff
@@ -1439,6 +1439,45 @@ async def run_engine(
...
+            local_work = self.context.get_active_request_count() + len(self.waiting_request_ids)
+
+        if parallel_state.get_expert_model_parallel_world_size() > 1:
+            expert_model_parallel_group = parallel_state.get_expert_model_parallel_group()
+            # all reduce local work across expert model parallel group
+
+            local_work_tensor = torch.tensor([local_work], device=torch.cuda.current_device())
+            torch.distributed.all_reduce(
+                local_work_tensor,
+                op=torch.distributed.ReduceOp.SUM,
```
Comment (lmcafee-nvidia) [inline megatron/core/inference/engines/dynamic_engine.py]: might be slightly clearer to the reader to use `MAX`, since all we care about is if there is a non-zero present.

File: megatron/core/inference/engines/dynamic_engine.py [hunk 5]
```diff
@@ -1462,26 +1501,35 @@ async def run_engine_with_coordinator(
                 # needed to send one message on an IPC socket. However
                 # just to be safe, we use 20ms here.
 
+                pending_requests = self.context.get_active_request_count() + len(
```
Comment (lmcafee-nvidia) [inline megatron/core/inference/engines/dynamic_engine.py]: nit, could just return a tuple of `(pending_requests, global_work > 0)` from `_ep_group_has_work()`, instead of computing `pending_requests` both here and (`local_work`) in `_ep_group_has_work()`

File: megatron/core/inference/engines/dynamic_engine.py [hunk 6]
```diff
@@ -1462,26 +1501,35 @@ async def run_engine_with_coordinator(
...
+                    self.waiting_request_ids
+                )
+                ep_group_has_work = self._ep_group_has_work()
+
+                if ep_group_has_work and pending_requests == 0:
+                    # run dummy forward pass if EP group as a whole has work,
+                    # but this rank does not have any work.
+                    self.controller.dummy_forward()
```
Comment (lmcafee-nvidia) [inline megatron/core/inference/engines/dynamic_engine.py:1516]: why do we need this dummy forward, instead of going straight to `continue`?

File: examples/inference/gpt/gpt_dynamic_inference_with_coordinator.py [hunk 7]
```diff
@@ -61,14 +62,6 @@ async def main(
         verbose=True,
     )
 
-    # if you want to use your own inference coordinator -
```
Comment (lmcafee-nvidia) [inline examples/inference/gpt/gpt_dynamic_inference_with_coordinator.py:63]: why was this removed?

File: megatron/core/inference/engines/dynamic_engine.py (hunk 6 above)
Comment (sidsingh-nvidia) [reply megatron/core/inference/engines/dynamic_engine.py:1516]: EP ranks also double up as DP ranks.
An EP rank might not have any work (example - the dp coordinator hasn't assigned it any work yet) but it's peers can have work.
We still need that EP rank to do a forward pass, because we require it's presence in the expert parallel All to All.
So we make it do a dummy forward.

File: examples/inference/gpt/gpt_dynamic_inference_with_coordinator.py (hunk 7 above)
Comment (sidsingh-nvidia) [reply examples/inference/gpt/gpt_dynamic_inference_with_coordinator.py:63]: I had this assuming that the RL folks will build their own coordinator. But with the latest merge they are basically using my original coordinator with some extra code for signal handling.

File: megatron/core/inference/engines/dynamic_engine.py (hunk 6 above)
Comment (lmcafee-nvidia) [reply megatron/core/inference/engines/dynamic_engine.py:1516]: ah, makes sense

**Comment by sidsingh-nvidia** [comment]
@kvareddy think this is ready for your pass. 
re:RL issues - Will not merge this until @jon-barker signs off.

**Review by kvareddy — APPROVED** [review]

File: megatron/core/inference/engines/dynamic_engine.py (hunk 2 above)
Comment (wdykas) [reply megatron/core/inference/engines/dynamic_engine.py]: yes

File: megatron/core/inference/engines/dynamic_engine.py [hunk 8]
```diff
@@ -1470,27 +1500,39 @@ async def run_engine_with_coordinator(
...
+                if ep_group_has_work and local_pending_requests == 0:
+                    # run dummy forward pass if EP group as a whole has work,
+                    # but this rank does not have any work.
+                    self.controller.dummy_forward()
                     continue
 
-                else:
-                    self.resume()
+                # 3. No work in EP group - Processing Control Signals
+                if not ep_group_has_work:
```
Comment (sidsingh-nvidia) [inline megatron/core/inference/engines/dynamic_engine.py:1528]: write comment here explaining why we are using the sync to process signals

File: megatron/core/inference/text_generation_controllers/text_generation_controller.py [hunk 9]
```diff
@@ -747,6 +747,26 @@ def _dynamic_step_calculate_top_n_logprobs(
...
+        # if no cuda graphs, directly use dummy forward
+        if not context.cuda_graph_batch_dimensions_list:
+            return self.inference_wrapped_model.dummy_forward()
+
+        # attempt to use cuda-graph if possible
+        input_ids, position_ids = self._dynamic_step_context_init(
+            # use the smallest cuda-graph config for dummy forward
+            construct_graph_dimensions=min(context.cuda_graph_batch_dimensions_list)
+        )
+        if context.using_cuda_graph_this_step():
```
Comment (sidsingh-nvidia) [inline megatron/core/inference/text_generation_controllers/text_generation_controller.py:775]: clarify this in comments

File: megatron/core/inference/engines/dynamic_engine.py [hunk 10]
```diff
@@ -1447,6 +1446,39 @@ async def run_engine(self, *, loop: Optional[asyncio.AbstractEventLoop] = None):
...
+            # is yet to receive the signal.
+            # So this is an *attempt* to process the signal. This rank has received the signal
+            # and passes 0 to the all-reduce. If any other rank in the EP group has not received the signal yet,
+            # it will pass a non-zero value to the all-reduce, and hence the global work will be non-zero,
+            # and we will defer processing the signal.
+            # When all ranks receive the signal, global work will be zero, and we can process the signal safely.
+            local_work = 0
+
+        if self.ep_world_size > 1:
+            max_global_work = await self.expert_parallel_zmq_communicator.all_reduce_max(local_work)
```
Comment (sidsingh-nvidia) [inline megatron/core/inference/engines/dynamic_engine.py:1478]: add why we are using this instead of torch.distributed.
